Tests/helper.py

```diff
@@ -126,7 +126,7 @@ def assert_image_similar(a, b, epsilon, msg=None):
         if HAS_UPLOADER:
             try:
                 url = test_image_results.upload(a, b)
-                logger.error(f"Url for test images: {url}")
+                logger.exception("URL for test images: %s", url)
```

Comment:
Why is this `logger.exception` but the earlier change in this file is still `logger.error`?

Reply:
Because it's inside an `except` block (see line 125), the nested `try` isn't relevant.

Compare:

```pycon
>>> import logging
>>>
>>> # both `error`, don't get any extra exception info:
>>> try:
...     logging.error(12)
...     raise ValueError
... except ValueError:
...     logging.error(34)
...
ERROR:root:12
ERROR:root:34
>>>
```
```pycon
>>> # `exception` in the `except` gives extra info:
>>> try:
...     logging.error(12)
...     raise ValueError
... except ValueError:
...     logging.exception(34)
...
ERROR:root:12
ERROR:root:34
Traceback (most recent call last):
  File "<stdin>", line 3, in <module>
ValueError
>>>
```
```pycon
>>> # `exception` in the `try` gives unhelpful info:
>>> try:
...     logging.exception(12)
...     raise ValueError
... except ValueError:
...     logging.error(34)
...
ERROR:root:12
NoneType: None
ERROR:root:34
>>>
```

Reply:
But line 94 and line 129 are both within `try` blocks with `except Exception`. I'm not seeing why one should be `logger.error` and the other `logger.exception`.

Reply:
It needs to be in the `except` part:

```python
try:
    logger.error(123)
except:
    logger.exception(456)  # <-- here
```

Only at 456 has an exception occurred, and there's exception info to be logged.

The first one only has the logger in the `try` block, not the `except` block:

```python
            try:
                url = test_image_results.upload(a, b)
                logger.error("URL for test images: %s", url)  # <-- no exception info here
            except Exception:
                pass  # <-- there's exception info here, if we wanted to log it
```

The second one:

```python
    try:
        assert epsilon >= ave_diff, (
            (msg or "")
            + f" average pixel value difference {ave_diff:.4f} > epsilon {epsilon:.4f}"
        )
    except Exception as e:
        if HAS_UPLOADER:
            try:
                url = test_image_results.upload(a, b)
                logger.exception("URL for test images: %s", url)  # <-- has exception info from the outer try/except
            except Exception:
                pass
        raise e
```

The `logger.exception` is in the `try` of the inner one, but that doesn't matter. What matters is it's in the `except` of the outer one.

Simplifying a bit, this is what matters:

```python
    try:
        ...
    except Exception as e:
        logger.exception("URL for test images: %s", url)  # <-- exception info here
```

Does that make sense?

Reply:
Ok, thanks. You were explaining it clearly, I just wasn't getting it.